paddle/fluid/memory/malloc.cc

```diff
@@ -28,12 +28,9 @@ namespace memory {
 using BuddyAllocator = detail::BuddyAllocator;
 
 BuddyAllocator* GetCPUBuddyAllocator() {
-  static detail::BuddyAllocator* a = nullptr;
-  if (a == nullptr) {
-    a = new detail::BuddyAllocator(new detail::CPUAllocator,
-                                   platform::CpuMinChunkSize(),
-                                   platform::CpuMaxChunkSize());
-  }
+  static detail::BuddyAllocator* a = new detail::BuddyAllocator(
```

Comment:
Not sure this is correct, use std::call_once could be more confident.

Reply:
I'm not sure that `new` will work. Because it's the pointer which is `static`, according to your link, the pointer is thread-safe though but I'm not sure that allocating a new object on heap can be.

Reply:
Why we need to new a CPUBuddyAllocator?

Just 

```cpp
static detail::BuddyAllocator a(new detail::CPUAllocator(), ...);
return &a;
```

is cool?

Reply:
我看了这个的解释之后，感觉咱们两个的写法可能都是不对的。

最简单的办法是用std::call_once

文档在这里：http://www.aristeia.com/Papers/DDJ_Jul_Aug_2004_revised.pdf

Reply:
这里有一个总结：http://hungmingwu-blog.logdown.com/posts/248195-singleton-implementation-iand-discussionn-c